### PR TITLE
[SYCL] Merge USMAllocContext for Sub-Devices and Sub-Sub-Devices

### PR DESCRIPTION
**Pull request by uditagarwal97**
**Problem:**
Currently, sub-devices and sub-sub-devices have different USM allocation contexts (USMAllocContext) but they share the same ze_device_handle. This causes issues with USM memory allocation, for instance, double-free.

**Proposed Fix:**
I have merged the USMAllocContext of the sub-device and sub-sub-device.

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -8195,7 +8229,7 @@ pi_result piextUSMDeviceAlloc(void **ResultPtr, pi_context Context,
     if (It == Context->DeviceMemAllocContexts.end())
       return PI_ERROR_INVALID_VALUE;
 
-    *ResultPtr = It->second.allocate(Size, Alignment);
+    *ResultPtr = It->second.get()->allocate(Size, Alignment);
```
Comment (aelovikov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:8232]: What is the type of `It->second`? Is that `std::shared_ptr`? If so, it should have `operator->` and we don't need to call `.get()`.

**Comment by uditagarwal97** [comment]
The issue is reproducible on the existing test (the one being updated in https://github.com/intel/llvm-test-suite/pull/1515).
We've only seen it internally because the project's CI doesn't use PVC machines currently.

**Comment by smaslov-intel** [comment]
> **Problem:** Currently, sub-devices and sub-sub-devices have different USM allocation contexts (USMAllocContext) but they share the same ze_device_handle. This causes issues with USM memory allocation, for instance, double-free.
> 
> **Proposed Fix:** I have merged the USMAllocContext of the sub-device and sub-sub-device.

should we just avoid freeing the context of sub-sub-devices since they do not carry unique L0 device handle?

Tag @igchor , the new owner of USM allocators

**Comment by igchor** [comment]
> **Problem:** Currently, sub-devices and sub-sub-devices have different USM allocation contexts (USMAllocContext) but they share the same ze_device_handle. This causes issues with USM memory allocation, for instance, double-free.
> 
> **Proposed Fix:** I have merged the USMAllocContext of the sub-device and sub-sub-device.

If the ze_device_handle is the same why not just use it as a key for `DeviceMemAllocContexts`, `SharedMemAllocContexts` and `SharedReadOnlyMemAllocContexts` instead of pi_device? This way you wouldn't have to create any shared ptrs - you could just call `insert` on `unordered_map` and if the key already exists, do nothing.

**Comment by smaslov-intel** [comment]
> DeviceMemAllocContexts

Sounds good to me!

**Comment by smaslov-intel** [comment]
> > **Problem:** Currently, sub-devices and sub-sub-devices have different USM allocation contexts (USMAllocContext) but they share the same ze_device_handle. This causes issues with USM memory allocation, for instance, double-free.
> > **Proposed Fix:** I have merged the USMAllocContext of the sub-device and sub-sub-device.
> 
> If the ze_device_handle is the same why not just use it as a key for `DeviceMemAllocContexts`, `SharedMemAllocContexts` and `SharedReadOnlyMemAllocContexts` instead of pi_device? This way you wouldn't have to create any shared ptrs - you could just call `insert` on `unordered_map` and if the key already exists, do nothing.

Sounds good to me!

**Comment by uditagarwal97** [comment]
Thanks for the feedback, @igchor and  @smaslov-intel. I have made the suggested changes.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -791,8 +793,12 @@ pi_result _pi_context::initialize() {
       [createUSMAllocators,
        &createUSMAllocatorsRecursive](pi_device Device) -> void {
     createUSMAllocators(Device);
-    for (auto &SubDevice : Device->SubDevices)
+    for (auto &SubDevice : Device->SubDevices) {
+      if (SubDevice->isCCS())
+      	// CCS share USMAllocContext with its root device.
+	continue;
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Sure, I've removed the intrinsic check and instead added an extra comment to explicitly state the sharing of USM allocators between CCS and subdevices.

**Review by smaslov-intel — APPROVED** [review]
LGTM

**Comment by igchor** [comment]
LGTM

**Comment by uditagarwal97** [comment]
Failure in SYCL :: dword_atomic_smoke.cpp is known #8098

**Comment by uditagarwal97** [comment]
@intel/llvm-gatekeepers The PR is ready!

**Pull request merged**